README.md
# World Flying Shooter
 Just a student having some fun with python.

Description: flying jet fighting different enemies.

Download files and assests and run 'Driver.py' (python Driver.py).
Dependecies: python, pygame, pygame_gui

Game: default arrows to move, default auto shoot, can be changed in main menu under settings. Game currently consists of 3 stages.
Each stage introduces a new enemy and consists of multiple levels, or parts. Get the required score to beat a part. Collect different buffs to help you out as you go. Player stats save between parts but not stages. If player runs out of health they die and reset the part. If they run out of lives. they reset the stage and their stats. Your player is saved into a profile that can later be loaded into the game.

Changes can be found in the changelog.txt and plans for next updates in roadwork.txt. You can also find more information about the game and future content in the documentation file.

Known bug: game might be bigger than your screen. Make sure that in yoru display settings, the scale is set to 100%. Any questions you can ask in my git profile.



changelog.txt
----------------------------------------------------------------------------------------------------------------------
                                               Change Log
----------------------------------------------------------------------------------------------------------------------
1.0: *fighting jet that has to dodge missiles.
     *sounds and music work.
----------------------------------------------------------------------------------------------------------------------
1.5: *added shooting and fixed shooting rate.
     *added health, damage and added score.
-----------------------------------------------------------------------------------------------------------------------
2.0: working menu using pygame_gui, created menu as driver class, game gets its own class, gets called in driver class.
-----------------------------------------------------------------------------------------------------------------------
2.1: *Added some text info in game.
     *organized classes and created folders
     *updated first level name and enemy.
----------------------------------------------------------------------------------------------------------------------
2.2: *menu fixed for true loop. worked mostly on a working retry menu. now menus are callable functions in driver.
     *Getting a score of 50 wins and kicks you out to main menu.
----------------------------------------------------------------------------------------------------------------------
2.3: *powerup works.
     *first bug: bullets start spawning from the left randomy, related to too fast maybe?
----------------------------------------------------------------------------------------------------------------------
2.4: *Added settings menu, user can be pick between auto and manual shooting. default to auto.
     *fixed shooting bug.
     *Created Menus class in Other folder for different menus.
----------------------------------------------------------------------------------------------------------------------
2.5: * fixed white background showing with sprites by using a different color on bg then use white, and save as png.
     * Added green jet enemy. More health and damage but gives 3 points if killed.
     * Split game code into two main loops: wave 1 and wave 2.
     * Created three different buff actors: health buff(+10), damage buff(+50), bullet per second buff(-50ms to timer).
     * every 20 points in wave 1, every 40 in wave 2: 50%: spawn health buff, 25%: spawn damage buff, 25%: bullet/s buff.
     * changed window size to 1920 * 1080, and made it fullscreen by default.
     * added display options in settings menu: fullscreen, hardware accelerated fullscreen, and window that's scalable.
     * added settings button to try again menu.
     * rescaled the menus, actor sprites, level text position (took longer than it sounds okay...).
     * fixed escape press bug that will jump the game to next wave (since it would end the first loop but not second).
     * added different ways to move: arrows(default), wads keys, and mouse.
     * added different ways to shoot: auto(default), space, and mouse.
     * added user selection on menu.
-----------------------------------------------------------------------------------------------------------------------
2.6: * slow day today... got a HUD working that can be used across levels
     * Got the first level working with the new HUD, took forever
     * First three levels working.
-----------------------------------------------------------------------------------------------------------------------
2.62:* got enemy shooting mechanics
     * got 4 levels now
     * fixed where buffs were spawning too high
     * fixed a enemy spawn issue
----------------------------------------------------------------------------------------------------------------------
2.65:* Restructured all of the levels for more organization and better control
     * Each part in the loop is well documented
     * Restructured buffs and enemies into their classes for easier calling
----------------------------------------------------------------------------------------------------------------------
2.68 * All this restructuring led to a level maker... YUP! I FEEL LIKE GOD! Level maker can take in arguments and
       generate a whole level without having to write the whole code for it... Still need to add more features like
       multiple enemies at the same time but holy shit i did it! Object oriented up the ass!
-----------------------------------------------------------------------------------------------------------------------
2.7 * what's a level maker without a level maker script? Yup that's done. pretty good for now just don't input wrong values!
      Now I just have to implement into back in my level maker and make a level unpacker. Hope it goes well or I oop!
-----------------------------------------------------------------------------------------------------------------------
2.8 * 100% done with implementing my level maker idea! Everything from the script maker works. Im not sure how accurate
      the random is but I honestly its amazing what I've done.
    * Created tabs in documentation for proper placement.
-----------------------------------------------------------------------------------------------------------------------
3.0 ** Welcome to 3.0 baby! A lot has changed since 1.0 version. We're now in the automation era.
    * Changes level style. So game consists of stages, and each stage has multiple parts or levels. So stages are
      packs of parts (levels). The first 3 stages are implemented and you can play them right now. The player info like
      health, damage, and bullet speeds transfer over between each part and reset when a stage is done. What about waves?
      Waves will still be implemented on harder stages so parts can still have waves.
    * Player now starts with full health. You lose if you run out of health. If you run out of lives, you reset the stage

Other/Menus.py
from Other.GuiHelper import GuiHelper
from Other.PlayerPanel import PlayerPanel
import pygame as py


# ====================================== MENUS FOR GAME ===============================================
# For buttons use 70 pixel space between buttons in height


class GameMenu:
    def __init__(self, width, height, manager):
        self.SW = width
        self.SH = height
        self.manager = manager

        # main menu
        self.play_b = None
        self.info_b = None
        self.quit_b = None

        # replay
        self.replay_b = None
        self.quit_replay_b = None

        # settings
        self.movement_selection = None
        self.screen_selection = None
        self.shoot_selection = None
        self.confirm_setting_b = None

        # next level
        self.next_level_b = None
        self.quit_next_level_b = None

        # profile select
        self.save_selection = None
        self.launch_hub_b = None
        self.quit_launcher_b = None

        # player hub
        self.cont_b = None
        self.store_b = None
        self.store2_b = None
        self.setting_b = None
        self.stats_b = None
        self.back_b = None
        self.player_panel = PlayerPanel(self.manager)
        self.pick_b = None
        self.pick_ddm = None

        # gui helper maker
        self.maker = GuiHelper(self.SW, self.SH, self.manager)
        self.SW2 = int(self.SW / 2)
        self.SH2 = int(self.SH / 2)

    def main_menu(self):
        # buttonS2 = (70, 60)

        title = self.maker.make_label(self.SW2, 200, 1000, 110, 'WORLD FLYING SHOOTER', 'maintitle', None)

        button_s = (150, 60)
        height = self.SH2 + 20

        self.play_b = self.maker.make_button(self.SW2, height, button_s[0], button_s[1], 'Play', None)

        self.info_b = self.maker.make_button(self.SW2, height + 70, button_s[0], button_s[1], 'How to Play', None)

        self.quit_b = self.maker.make_button(self.SW2, height + (70 * 2), button_s[0], button_s[1], 'Quit', None)

    def replay_menu(self, string):
        title = self.maker.make_label(self.SW2, self.SH2 - 90, 260, 90, string, 'replay', None)

        button_s = (150, 60)

        self.replay_b = self.maker.make_button(self.SW2, self.SH2 + 100, button_s[0], button_s[1], 'play again',
                                               None)

        self.quit_replay_b = self.maker.make_button(self.SW2, self.SH2 + 200, button_s[0], button_s[1],
                                                    'back to hub',
                                                    None)

    def settings_menu(self):
        button_s = (150, 60)

        self.movement_selection = self.maker.make_selection_list(self.SW2, self.SH2 - 70, 200, 70,
                                                                 ['arrow keys', "wads keys", "mouse"], 'movementS',
                                                                 None)

        self.screen_selection = self.maker.make_selection_list(self.SW2, self.SH2 + 20, 200, 70,
                                                               ["fullscreen", "hardware accelerated",
                                                                "windowed"], 'screenS', None)

        self.shoot_selection = self.maker.make_selection_list(self.SW2, self.SH2 + 110, 200, 70,
                                                              ['auto', "space", "mouse"], 'shootS', None)

        self.confirm_setting_b = self.maker.make_button(self.SW2, self.SH2 + 300, button_s[0], button_s[1], 'exit'
                                                        , None)

    def next_level(self, string):
        label_size = (240, 90)
        button_s = (150, 60)

        title = self.maker.make_label(self.SW2, self.SH2 - 90, label_size[0], label_size[1], string, 'nextlevel',
                                      None)

        self.next_level_b = self.maker.make_button(self.SW2, self.SH2 + 100, button_s[0], button_s[1], string
                                                   , None)

        self.quit_next_level_b = self.maker.make_button(self.SW2, self.SH2 + 200, button_s[0], button_s[1],
                                                        'back to hub'
                                                        , None)

    # pick between file saves
    def load_menu(self, names):
        todo = ''
        # labelSize = (0, 0, 150, 40)
        # label_rect = pygame.Rect(labelSize)
        #
        # label_rect.bottomright = (int(- self.SW / 2) + 20, -700)
        # label1 = pygame_gui.elements.UILabel(relative_rect=label_rect, text='Pick a save',
        #                                      manager=self.manager, anchors=anchd, object_id='saveFL')
        #
        # shootDSize = (0, 0, 200, 150)
        # dD_rect = pygame.Rect(shootDSize)
        #
        # dD_rect.bottomright = (int(- self.SW / 2 + 50), -520)
        # self.savelist_selection = pygame_gui.elements.UISelectionList(dD_rect, names, object_id='loadS',
        #                                                         manager =self.manager, allow_multi_select=False,
        #                                                               anchors=anchd)
        #
        # buttonSize = (0, 0, 150, 60)
        # button_layout_rect = pygame.Rect(buttonSize)
        #
        # button_layout_rect.bottomright = (int(-self.SW / 2) + 20, -400)
        # self.loadConfirm_button = pygame_gui.elements.UIButton(relative_rect=button_layout_rect,
        #                                             text='Confirm', manager=self.manager,
        #                                             anchors=anchd)

    # save menu
    def save_menu(self, names):
        todo = ''

        # labelSize = (0, 0, 150, 40)
        # label_rect = pygame.Rect(labelSize)
        # labesize2 = (0,0,210,40)
        #
        # label_rect.bottomright = (int(- self.SW / 2) + 25, -700)
        # label1 = pygame_gui.elements.UILabel(relative_rect=label_rect, text='Pick a save',
        #                                     manager=self.manager, anchors=anchd, object_id='saveFL')
        #
        # shootDSize = (0, 0, 200, 150)
        # dD_rect = pygame.Rect(shootDSize)
        #
        # dD_rect.bottomright = (int(- self.SW / 2 + 50), -500)
        # self.savelist2_selection = pygame_gui.elements.UISelectionList(dD_rect, names, object_id= 'saveS',
        #                                                         manager= self.manager, allow_multi_select=False,
        #                                                                anchors=anchd)
        #
        # label_rect = pygame.Rect(labesize2)
        # label_rect.bottomright = (int(- self.SW / 2 + 60), -400)
        # label2 = pygame_gui.elements.UILabel(relative_rect=label_rect, text='Or new save',
        #                                      manager=self.manager, anchors=anchd, object_id='saveFL2')
        #
        # buttonSize = (0, 0, 150, 60)
        # button_layout_rect = pygame.Rect(buttonSize)
        #
        # button_layout_rect.bottomright = (int(-self.SW / 2) - 100, -170)
        # self.saveConfirm_button = pygame_gui.elements.UIButton(relative_rect=button_layout_rect,
        #                                             text='Confirm', manager=self.manager,
        #                                             anchors=anchd)
        #
        # button_layout_rect.bottomright = (int(-self.SW / 2) + 100, -170)
        # self.cancel_button = pygame_gui.elements.UIButton(relative_rect=button_layout_rect,
        #                                                        text='Cancel', manager=self.manager,
        #                                                        anchors=anchd)

    # description menu TODO
    def info_menu(self):
        todo = ''
        # labelSize = (0, 0, 500, 120)
        # label_rect = pygame.Rect(labelSize)
        #
        # label_rect.bottomright = (int(- self.SW / 2 + 200), -600)
        # title = pygame_gui.elements.UILabel(relative_rect=label_rect, text='no exit',
        #                                     manager=self.manager, anchors=anchd)

    # get player save or make a new one
    def launch_menu(self, names):
        label1 = self.maker.make_label(self.SW2, self.SH2 - 170, 210, 40, 'Load a Save', 'l1', None)

        self.save_selection = self.maker.make_selection_list(self.SW2, self.SH2, 200, 150, names, 'saveS', None)

        label2 = self.maker.make_label(self.SW2, self.SH2 + 120, 210, 40, 'Or New Profile', 'l2', None)

        self.launch_hub_b = self.maker.make_button(self.SW2 + 80, self.SH2 + 300, 150, 60, 'Go', None)
        self.quit_launcher_b = self.maker.make_button(self.SW2 - 80, self.SH2 + 300, 150, 60, 'Back', None)

    # player hub
    def player_hub(self, player_save, stages):
        # menu buttons
        self.cont_b = self.maker.make_button(400, 330, 150, 60, 'continue', None)
        self.pick_b = self.maker.make_button(350, 400, 50, 60, 'Go', None)
        self.pick_ddm = self.maker.make_drop_down_menu(425, 400, 100, 60, 'pick_ddm', stages)
        self.store_b = self.maker.make_button(400, 470, 150, 60, 'store', None)
        self.stats_b = self.maker.make_button(400, 610, 150, 60, 'stats', None)
        self.setting_b = self.maker.make_button(400, 680, 150, 60, 'settings', None)
        self.back_b = self.maker.make_button(400, 750, 150, 60, 'save and exit', None)

        # player info panel
        self.player_panel.setPlayer(player_save)
        #  ====================================== player name panel ==============================================
        name_panel = self.maker.make_panel(self.SW2 - 100, 100, 400, 180, 'namepanel')

        # title
        string = "Welcome Back " + str(player_save['name'])
        player_name = self.maker.make_label(200, 30, 300, 30, string, 'playername', name_panel)

        #  --------------------------------------------- times -----------------------------------------------------
        times = player_save['times']
        string = 'Created: ' + str(times[0])
        dob_label = self.maker.make_label(200, 60, 260, 20, string, 'dobL', name_panel)

        string = 'Last played: ' + str(times[1])
        last_save_label = self.maker.make_label(200, 90, 260, 20, string, 'time2L', name_panel)

        player = player_save['player']
        total_time = self.maker.make_label(200, 120, 260, 20, ('total game time: ' + str(player['time']) + 's'),
                                                'gametime', name_panel)